### PR TITLE
NIFI-7386: Azurite emulator support

### PR DESCRIPTION
**Pull request by sjyang18**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [X] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by sjyang18** [comment]
@esecules, @jfrazee This is the approach I suggest we go to support Azurite emulator.

**Comment by sjyang18** [comment]
reference documentation in implementing the support for azurite emulator. 

https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string#connect-to-the-emulator-account-using-a-shortcut

https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string#specify-an-http-proxy

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/storage/AzureStorageEmulatorCrendentialsControllerService.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ */
+@Tags({ "azure", "microsoft", "emulator", "storage", "blob", "queue", "credentials" })
+@CapabilityDescription("Defines credentials for Azure Storage processors that connects to Azurite emulator. ")
+public class AzureStorageEmulatorCrendentialsControllerService extends AbstractControllerService implements AzureStorageCredentialsService {
+
+
+    public static final PropertyDescriptor DEVELOPMENT_STORAGE_PROXY_URI = new PropertyDescriptor.Builder()
+            .name("azurite-proxy-uri")
+            .displayName("Azurite Proxy URI")
+            .description("Default null will connect to http://127.0.0.1. Otherwise, overwrite this value with your proxy url.")
```
[the remark this answers is not in the document]
Comment (sjyang18) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/storage/AzureStorageEmulatorCrendentialsControllerService.java]: changed as you suggested.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/storage/AzureStorageEmulatorCrendentialsControllerService.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ * @see AbstractControllerService
+ */
+@Tags({ "azure", "microsoft", "emulator", "storage", "blob", "queue", "credentials" })
+@CapabilityDescription("Defines credentials for Azure Storage processors that connects to Azurite emulator. ")
+public class AzureStorageEmulatorCrendentialsControllerService extends AbstractControllerService implements AzureStorageCredentialsService {
+
+
+    public static final PropertyDescriptor DEVELOPMENT_STORAGE_PROXY_URI = new PropertyDescriptor.Builder()
+            .name("azurite-proxy-uri")
+            .displayName("Azurite Proxy URI")
```
[the remark this answers is not in the document]
Comment (sjyang18) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/storage/AzureStorageEmulatorCrendentialsControllerService.java]: changed as you suggested.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/services/azure/storage/TestAzureStorageEmulatorCredentialsControllerService.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+package org.apache.nifi.services.azure.storage;
+
+import org.apache.nifi.reporting.InitializationException;
+import org.apache.nifi.util.NoOpProcessor;
+import org.apache.nifi.util.TestRunner;
+import org.apache.nifi.util.TestRunners;
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestAzureStorageEmulatorCredentialsControllerService {
```
Comment (esecules) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/test/java/org/apache/nifi/services/azure/storage/TestAzureStorageEmulatorCredentialsControllerService.java:26]: Does the test environment for the NiFi package support making integration tests where you actually spin up an instance of azurite and try to connect to it?

**Comment by esecules** [comment]
LGTM! what about you, @jfrazee ?

**Comment by esecules** [comment]
Bump! Having this feature will cut down on how much my test suite wait for network I/O.

**Comment by sjyang18** [comment]
I rebased my branch on top of main and resolved the conflict. @jfrazee, Would you take a look?

**Review by jfrazee — CHANGES_REQUESTED** [review]
There are a handful of minor comments but one bigger thing: I successfully tested this with Azurite and Blob but it doesn't work with Queues because we need to add the handling for the two types of credentials into [`createCloudQueueClient`](https://github.com/apache/nifi/blob/47b167482d8df5d2722fa86d808a787418159f9f/nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/queue/AbstractAzureQueueStorage.java#L69-L81) in `AbstractAzureQueueStorage`.

BTW, while working through this I kept wondering whether this should really be specifically for the emulator or whether we just needed support [connection strings](https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string). I decided it makes sense to proceed with this but opened [NIFI-8097](https://issues.apache.org/jira/browse/NIFI-8097) to add this functionality.

**Review by jfrazee — COMMENTED** [review]
@sjyang18 +1 Thanks for this.

**Pull request closed without merging**